src/effects/VST3/VST3Wrapper.cpp

```diff
@@ -630,3 +630,26 @@ void VST3Wrapper::SaveUserPreset(const EffectDefinitionInterface& effect, const
          SetConfig(effect, PluginSettings::Private, name, controllerStateKey, *vst3settings.controllerState);
    }
 }
+
+void VST3Wrapper::AssignSettings(EffectSettings& dst, EffectSettings&& src)
+{
+   if(dst.cast<VST3EffectSettings>() == nullptr)
+      dst = src;
+}
+
+void VST3Wrapper::CopySettingsContents(const EffectSettings& src, EffectSettings& dst, SettingsCopyDirection copyDirection)
+{
+   if(copyDirection == SettingsCopyDirection::MainToWorker)
+   {
+      //Clear changes that were already applied, so they won't appear
+      //on the other side on swap.
+      GetSettings(dst).parameterChanges.clear();
+      //Don't allocate in worker
+      std::swap(
+         GetSettings(*const_cast<EffectSettings*>(&src)).parameterChanges,
+         GetSettings(dst).parameterChanges
+      );
+   }
+   else
+      GetSettings(dst).parameterChanges.clear();
```

Comment:
Moving on the the ninth commit.  As I already mentioned, you should avoid either allocating or freeing and that means, not clearing a `std::map` either.  So I suggest again, a map from keys to optional values and just reassignment of null opt instead of clearing.

Reply:
`CopySettingsContents` in either direction is always done on the worker thread.